modules/nf-commons/src/main/nextflow/plugin/OciAwareFileDownloader.groovy

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2013-2025, Seqera Labs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package nextflow.plugin
+
+import groovy.transform.CompileStatic
+import groovy.util.logging.Slf4j
+import org.pf4j.update.SimpleFileDownloader
+
+import java.nio.file.Files
+import java.nio.file.Path
+import java.util.regex.Pattern
+
+/**
+ * FileDownloader extension that enables the download of OCI compliant artifact that require a token authorization.
+ *
+ * @author Jorge Ejarque <jorge.ejarque@seqera.io>
+ */
+@Slf4j
+@CompileStatic
+class OciAwareFileDownloader extends SimpleFileDownloader {
+
+    private static final Pattern WWW_AUTH_PATTERN = ~/Bearer realm="([^"]+)",\s*service="([^"]+)",\s*scope="([^"]+)"/
+
+    /**
+     * OCI aware download with token authorization. Tries to download the artifact and if it fails checks the headers to get the
+     * @param fileUrl source file
+     * @return
+     */
+    @Override
+    protected Path downloadFileHttp(URL fileUrl) {
+
+        Path destination = Files.createTempDirectory("pf4j-update-downloader");
+        destination.toFile().deleteOnExit();
+
+        String path = fileUrl.getPath();
+        String fileName = path.substring(path.lastIndexOf('/') + 1);
+        Path file = destination.resolve(fileName);
+        HttpURLConnection conn = (HttpURLConnection) fileUrl.openConnection()
+        conn.instanceFollowRedirects = true
+
+        if (conn.responseCode == HttpURLConnection.HTTP_UNAUTHORIZED) {
+            def wwwAuth = conn.getHeaderField("WWW-Authenticate")
+            if (wwwAuth?.contains("Bearer")) {
```

Comment:
Note that this may also be `Basic`. It may be worth considering adopting the ORAS client to avoid to maintain all this burden. My initial assumption was to avoid that because the registry it was expected to not require any auth.